api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -506,6 +507,51 @@ message HttpConnectionManager {
   //
   // 3. Tracing decision (sampled, forced, etc) is set in 14th byte of the UUID.
   RequestIDExtension request_id_extension = 36;
+
+  // Configuration of local reply returned by Envoy. If Envoy needs to reject a request due to some
+  // error conditions, it calls sendLocalRely with a http status code and detail error text. If this
+  // config is not specified, the error text will be send as response body in plain text format for
+  // http, in grpc-message header for grpc. This config supports rewritting the status code and error
+  // text. The new error text can be combined from many access log operators, such as %REP(PATH)%.
+  // Its format can be either plain text or JSON. The new error text will be send as response body in
+  // either plain/text or application/json format for http, and in grpc-message header for grpc.
+  LocalReplyConfig local_reply_config = 38;
+}
+
+message LocalReplyConfig {
+  // Configuration of list of mappers which allows to filter and change local response.
+  // The client will iterate through mappers until first match.
+  repeated ResponseMapper mapper = 1;
+
+  // Specify how to form a new text body from access log :ref:`command operators <config_access_log_command_operators>`
+  // and in which format: either plain text or JSON. This is the catch-all format, it will be used
+  // if the matched mapper doesn't have its own format specified or there is not any mappers.
+  // If not specified, the text body will not be changed, the error detail passed in the
+  // sencLocalReply call will be used. The format will be plain text.
+  type.v3.StringOrJson format = 2;
+}
+
+message ResponseMapper {
+  // Filter is used to determine if the response should be rewritten.
+  config.accesslog.v3.AccessLogFilter filter = 1 [(validate.rules).message = {required: true}];
+
+  // Rewriter defines how to rewrite response.
+  ResponseRewriter rewriter = 2 [(validate.rules).message = {required: true}];
+
+  // Specify how to form a new text body from access log :ref:`command operators <config_access_log_command_operators>`
+  // and in which format: either plain text or JSON. It is used only when this mapper is matched.
+  // If not specified, the catch-all :ref:`format <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.LocalReplyConfig.format>`
+  // will be used.
+  type.v3.StringOrJson format = 3;
+}
+
+// Configuration of new value for matched local response.
+message ResponseRewriter {
+  // The new status code for matched response.
+  google.protobuf.UInt32Value status_code = 1 [(validate.rules).uint32 = {lt: 600 gte: 300}];
+
+  // The new response body for matched response.
+  google.protobuf.StringValue body = 2;
```

Comment:
I thought we had a type that wraps a oneof with either a string or bytes, like this:
https://github.com/envoyproxy/envoy/blob/e6e37da3d9854410cfc2b46f193b265563ee5580/api/envoy/data/tap/v3/common.proto#L20-L29

Can we add a core type that wraps ^? This is more flexible from a config perspective. @kyessenov can you remind us if this already exists/

Reply:
using "bytes" to support Unicode?

Reply:
Or any other raw bytes. This has come up a lot of times that it's easier to give people the option of how they want to represent body.

Reply:
How about using DataSource?  Some people may like to read it from local file too.

Reply:
Yes, sorry, that's what we should do. I think DataSource already has this type of oneof in it. Thank you!